File: action/driver_odrive/src/test_odrive_control.cpp
#include "rclcpp/rclcpp.hpp"
#include "can_msgs/msg/frame.hpp"
#include <iostream>
#include <chrono>
#include <map>
#include <vector>
#include <dbcppp/Network.h>
#include <geometry_msgs/msg/twist.hpp>
#include <fstream>

class CanPubSub : public rclcpp::Node {
public:
    CanPubSub() : Node("can_pubsub") {
        rclcpp::QoS custom_qos_profile(10);
        custom_qos_profile.avoid_ros_namespace_conventions(true);

        publisher_ = this->create_publisher<can_msgs::msg::Frame>("can_topic_pub", custom_qos_profile);
        subscription_ = this->create_subscription<can_msgs::msg::Frame>("can_topic_sub", custom_qos_profile,
            std::bind(&CanPubSub::topic_callback, this, std::placeholders::_1));

        control_subscription_ = this->create_subscription<geometry_msgs::msg::Twist>(
            "topic_name", 10, std::bind(&CanPubSub::control_callback, this, std::placeholders::_1));

        // Initialize variables
        current_time = std::chrono::system_clock::now();
        last_msg_time = std::chrono::system_clock::now();
        axis_list = {0, 1, 2, 3}; // Example axis list
        active_state = std::vector<int>(axis_list.size(), 0);
        buffer = std::vector<std::time_t>(axis_list.size(), 0);
        

    }

        
private:

    void control_callback(const geometry_msgs::msg::Twist::SharedPtr msg) const
    {
        RCLCPP_INFO(this->get_logger(), "Received twist message: Linear: x=%f, y=%f, z=%f Angular: x=%f, y=%f, z=%f",
                    msg->linear.x, msg->linear.y, msg->linear.z, msg->angular.x, msg->angular.y, msg->angular.z);
    }


    void topic_callback(const can_msgs::msg::Frame::SharedPtr msg) {
        current_time = std::chrono::system_clock::now();
        auto message = can_msgs::msg::Frame();

        for (int i : axis_list) {
            if (msg->id == (static_cast<unsigned int>(i) << 5 | 0x01)) {
                int errorCode = msg->data[0] | (msg->data[1] << 8) | (msg->data[2] << 16) | (msg->data[3] << 24);
                if (errorCode != 0) {
                    RCLCPP_ERROR(this->get_logger(), "Driver axis %d error code: %x", i, errorCode);
                    active_state[i] = 1;
                }
            }

            if (std::chrono::duration_cast<std::chrono::seconds>(current_time - std::chrono::system_clock::from_time_t(buffer[i])).count() > 1) {
                active_state[i] = 1;
            } else {
                if (active_state[i] == 1) {
                    active_state[i] = 0;
                    // Assuming db and bus are correctly defined and accessible

                    // const dbcppp::IMessage* dbcppp_msg = net->getMessageById(14);
                    // dbcppp::ISignal::signal_data_t data = 0;
                    // message.id = (0x18 | (i << 5));
                    // message.data = data;
                    // publisher_->publish(message);


                    // const dbcppp::IMessage* dbcppp_msg = net->getMessageById(7);
                    // dbcppp::ISignal::signal_data_t data = 0;
                    // dbcppp_msg->encode(data, "Axis_Requested_State", 0x08);
                    // message.id = (0x07 | (i << 5));
                    // message.data = data;
                    // publisher_->publish(message);

                    RCLCPP_INFO(this->get_logger(), "Clearing errors and setting axis state for axis %d", i);
                }
            }
        }

        if ((msg->id & 0x1F) == 0x09) {
            int axis_id = msg->id >> 5;
            buffer[axis_id] = std::chrono::system_clock::to_time_t(current_time);

            // Assuming db and bus are correctly defined and accessible
            // auto data = db.decode_message("Get_Encoder_Estimates", msg->data);
            // float v_cur = data["Vel_Estimate"];

            if (std::chrono::duration_cast<std::chrono::seconds>(current_time - last_msg_time).count() > 1) {
                wheel = 0;
            } else {
                if (axis_id % 2 == 0) {
                    wheel = vel_lin_setpoint - vel_ang_setpoint * l_distance / 2;
                } else {
                    wheel = vel_lin_setpoint + vel_ang_setpoint * l_distance / 2;
                }
            }
            

            //  // Create a message by ID - assuming 13 is the ID for Set_Input_Vel
            // const dbcppp::IMessage* dbcppp_msg = net->getMessageById(13);
            // dbcppp::ISignal::signal_data_t data = 0;
            // uint64_t input_vel_value = 2;      // Example value for Input_Vel
            // uint64_t input_torque_ff_value = 0; // Example value for Input_Torque_FF
            // dbcppp_msg->encode(data, "Input_Vel", input_vel_value);
            // dbcppp_msg->encode(data, "Input_Torque_FF", input_torque_ff_value);

            message.id = (axis_id << 5 | 0x00D);
            // message.data = data;
            publisher_->publish(message);

            if (DEBUG) {
                RCLCPP_INFO(this->get_logger(), "Received CAN Frame ID: '%d'", msg->id);
            }
        }

    }
    
    double l_distance = 1;
    double vel_lin_setpoint = 0;
    double vel_ang_setpoint = 0;
    double wheel = 0;
    bool arm_msg = true;
    const bool DEBUG = false;
    std::vector<int> axis_list;
    std::vector<int> active_state;
    std::vector<std::time_t> buffer;
    std::chrono::time_point<std::chrono::system_clock> current_time, last_msg_time;
    rclcpp::Subscription<can_msgs::msg::Frame>::SharedPtr subscription_;
    rclcpp::Subscription<geometry_msgs::msg::Twist>::SharedPtr control_subscription_;
    rclcpp::Publisher<can_msgs::msg::Frame>::SharedPtr publisher_;

};

int main(int argc, char* argv[]) {

    std::unique_ptr<dbcppp::INetwork> net;
    {
        std::ifstream idbc( "/home/rover/gr_platform2/colcon_ws/src/control/driver_odrive/config/odrive-cansimple.dbc");
        net = dbcppp::INetwork::LoadDBCFromIs(idbc);
    }

    std::unordered_map<uint64_t, const dbcppp::IMessage*> messages;
    for (const dbcppp::IMessage& msg : net->Messages())
    {   
        std::cout << "Hello" << "\n";
        messages.insert(std::make_pair(msg.Id(), &msg));
    }
    
    // Example: Set Input Velocity
    // auto set_input_vel = network->getMessageByName("Set_Input_Vel");
    // auto input_vel_signal = set_input_vel->getSignalByName("Input_Vel");
    // std::vector<uint8_t> data(set_input_vel->getMessageSize(), 0);
    // input_vel_signal->encode(data.data(), 1000.0);  // Set velocity to 1000 units

    // // Example: Clear Errors
    // auto clear_errors = network->getMessageByName("Clear_Errors");
    // // Assuming no signals need to be set for clearing errors

    // // Example: Set Axis State
    // auto set_axis_state = network->getMessageByName("Set_Axis_State");
    // auto axis_state_signal = set_axis_state->getSignalByName("Axis_Requested_State");
    // axis_state_signal->encode(data.data(), 1);  // Set state to 1 (example state)

    // // Example: Get Encoder Estimates
    // auto get_encoder_estimates = network->getMessageByName("Get_Encoder_Estimates");
    // auto vel_estimate_signal = get_encoder_estimates->getSignalByName("Vel_Estimate");
    // auto pos_estimate_signal = get_encoder_estimates->getSignalByName("Pos_Estimate");
    // // Assuming you need to read these values from a response

    // // Print encoded data for demonstration
    // std::cout << "Set Input Vel Data: ";
    // for (auto byte : data) {
    //     std::cout << std::to_string(byte) << " ";
    // }
    // std::cout << std::endl;
    
    // rclcpp::init(argc, argv);
    // auto node = std::make_shared<CanPubSub>();
    // rclcpp::spin(node);
    // rclcpp::shutdown();
    return 0;
}

File: action/driver_odrive/src/test_odrive_control.py
#!/usr/bin/python3

import rclpy
from rclpy.node import Node
from geometry_msgs.msg import Twist
from sensor_msgs.msg import Imu
from mavros_msgs.msg import VFR_HUD
from std_msgs.msg import Bool
import can
import cantools
import time
import numpy as np
import struct

class ControlNode(Node):
    def __init__(self):
        super().__init__('control_node')
        self.declare_parameters(namespace='', parameters=[
            ('dbc', None),
        ])
        dbc_file = self.get_parameter('dbc').get_parameter_value().string_value
        self.db = cantools.database.load_file(dbc_file)
        self.bus = can.Bus("can0", bustype="socketcan")

        self.bus.set_filters([
            {"can_mask": 0xFFF, "can_id": 0x01 | (0 << 5), "extended": False},
            {"can_mask": 0xFFF, "can_id": 0x01 | (1 << 5), "extended": False},
            {"can_mask": 0xFFF, "can_id": 0x01 | (2 << 5), "extended": False},
            {"can_mask": 0xFFF, "can_id": 0x01 | (3 << 5), "extended": False},
            
            {"can_mask": 0xFFF, "can_id": 0x09 | (0 << 5), "extended": False},
            {"can_mask": 0xFFF, "can_id": 0x09 | (1 << 5), "extended": False},
            {"can_mask": 0xFFF, "can_id": 0x09 | (2 << 5), "extended": False},
            {"can_mask": 0xFFF, "can_id": 0x09 | (3 << 5), "extended": False},
        ])

        self.vel_sub = self.create_subscription(Twist, 'diff_cont/cmd_vel_unstamped', self.cmd_vel_callback, 10)
        self.disarm_sub = self.create_subscription(Bool, 'cmd_trig', self.arm_callback, 10)
        # Uncomment and modify other subscribers as needed

        self.axis_list = [0, 1, 2, 3]
        self.control_dir = [1, 1, 1, -1]
        self.active_state = [1, 1, 1, 1]
        times = time.time()
        self.buffer = [times,times,times,times] 
        self.l_distance = 0.6
        self.last_msg_time = self.get_clock().now().to_sec()
        self.arm_msg = False
        self.ardupilot_vel_x = 0.0
        self.ardupilot_ang_z = 0.0
        self.vel_lin_setpoint = 0.0
        self.vel_ang_setpoint = 0.0
        self.DEBUG = False

    def arm_callback(self, msg):
        self.arm_msg = msg.data

    def cmd_vel_callback(self, msg):
        self.last_msg_time = self.get_clock().now().to_sec()
        self.vel_lin_setpoint = 1.2 * np.clip(msg.linear.x * 1.5, -4.0, 4.0)
        self.vel_ang_setpoint = 5.0 * np.clip(msg.angular.z * 1.5, -4.0, 4.0)

    def main_loop(self):

        for msg in self.bus:
            # Check the ros alive
            if rclpy.is_shutdown():
                print("bye")
                break
            
            if self.arm_msg:
                for i in self.axis_list:

                    # For each axis, check the error. Activate to clean and start the axis if detected
                    if(msg.arbitration_id == (i << 5 | 0x01)):
                        errorCode = msg.data[0] | msg.data[1] << 8 | msg.data[2] << 16 | msg.data[3] << 24
                        if not errorCode == 0x0:
                            print(f"Driver axis {i} error code: {str(hex(errorCode))}")
                            self.active_state[i] = 1
                    
                    # For each axis, check the timeout. Activate to clean and start the axis if timeout
                    if (time.time()-self.buffer[i]) > 1:
                        self.active_state[i] = 1
                    else:
                        if self.active_state[i] == 1:
                            self.active_state[i] = 0
                            data = self.db.encode_message('Clear_Errors', {})
                            self.bus.send(can.Message(arbitration_id=(0x18 | (i << 5)), is_extended_id=False, data=data))
                            data = self.db.encode_message('Set_Axis_State', {'Axis_Requested_State': 0x08})
                            self.bus.send(can.Message(arbitration_id=(0x07 | (i << 5)), is_extended_id=False, data=data))

                if (msg.arbitration_id & 0x1F) == 0x09:

                    axis_id = msg.arbitration_id >> 5
                    self.buffer[axis_id]=time.time()

                    # err_vel = 1.0*(vel_lin_setpoint - ardupilot_vel_x)
                    # err_ang = 1.0*(vel_ang_setpoint - ardupilot_ang_z)

                    data = self.db.decode_message('Get_Encoder_Estimates', msg.data)
                    v_cur = data['Vel_Estimate']

                    if time.time() - self.last_msg_time > 1:
                        wheel = 0
                    else: 
                        if axis_id%2 == 0:
                            wheel  = self.vel_lin_setpoint - self.vel_ang_setpoint*self.l_distance/2
                        else:
                            wheel = self.vel_lin_setpoint + self.vel_ang_setpoint*self.l_distance/2

                    self.bus.send(can.Message(arbitration_id=(axis_id << 5 | 0x00D), data=struct.pack('<ff', wheel*self.control_dir[axis_id], 0.0), is_extended_id=False))

                    if self.DEBUG:
                        if axis_id == 0:
                            # print(f"{round(err_vel,3)} {round(err_ang,3)}")
                            print(f"{round(self.vel_lin_setpoint,3)} {round(self.vel_ang_setpoint,3)}")

            time.sleep(0.004) # Set the delay for reducing the computational resources

        payload = self.db.encode_message('Set_Input_Torque', {'Input_Torque': 0.0})
        for axis_id in self.axis_list:
            msg = can.Message(arbitration_id=((axis_id << 5) | 0x00E), data=payload, is_extended_id=False)
            self.bus.send(msg)
        payload = self.db.encode_message('Set_Axis_State', {'Axis_Requested_State': 0x01})
        for axis_id in self.axis_list:
            msg = can.Message(arbitration_id=((axis_id << 5) | 0x007), data=payload, is_extended_id=False)
            self.bus.send(msg)


def main(args=None):
    rclpy.init(args=args)
    control_node = ControlNode()
    try:
        control_node.main_loop()
    except KeyboardInterrupt:
        pass
    finally:
        control_node.destroy_node()
        rclpy.shutdown()

if __name__ == '__main__':
    main()
